opm/simulators/flow/partitionCells.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+  Copyright 2021 Total SE
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <opm/simulators/flow/partitionCells.hpp>
+
+#include <fmt/format.h>
+
+#include <algorithm>
+#include <cassert>
+#include <fstream>
+#include <iterator>
+#include <numeric>
+#include <string>
+
+namespace Opm
+{
+
+
+    // Read from file, containing one number per cell.
+    std::pair<std::vector<int>, int> partitionCellsFromFile(const std::string& filename, const int num_cells)
+    {
+        // Read file into single vector.
+        std::ifstream is(filename);
+        const std::vector<int> cellpart{std::istream_iterator<int>(is), std::istream_iterator<int>()};
+        if (cellpart.size() != size_t(num_cells)) {
+            auto msg = fmt::format("Partition file contains {} entries, but there are {} cells.",
+                                   cellpart.size(), num_cells);
+            throw std::runtime_error(msg);
+        }
+
+        // Create and return the output domain vector.
+        const int num_domains = (*std::max_element(cellpart.begin(), cellpart.end())) + 1;
+        return { cellpart, num_domains };
+    }
+
+
+    // Trivially simple partitioner
+    std::pair<std::vector<int>, int> partitionCellsSimple(const int num_cells, const int num_domains)
+    {
+        // Build the partitions.
+        std::vector<int> bounds(num_domains + 1);
+        bounds[0] = 0;
+        const int dom_sz = num_cells / num_domains;
```

Comment:
What happens if `num_cells % num_domains != 0` here?

Reply:
The final domain will get the leftover cells: `bounds[num_domains-1] == dom_sz * (num_domains-1)`, but `bounds[num_domains] == num_cells`.

Reply:
>> What happens if `num_cells % num_domains != 0` here?
>
> The final domain will get the leftover cells

Right, but that's potentially a rather imbalanced partitioning, especially for the academic case of a small number of cells and a large number of domains.  For instance, if `num_cells = 399` and `num_domains = 100`, you're going to get 99 domains of size 3 and one domain of size 102.

Reply:
Indeed, but this is just a fallback and/or last resort, so I am not too worried about this. Also, for NLDD methods, domain imbalance is not necessary bad in itself: the main point is to localize iterations and that in itself does not require balanced domain sizes. I suggest we leave it rather than make something more complicated.

Reply:
> I suggest we leave it rather than make something more complicated.

I guess that's fine, but just for the record, it's not particularly complicated to create balanced bin sizes:
```C++
auto bounds  = std::vector<int>(num_domains + 1, dom_size);
bounds[0] = 0;
for (i = 0; i < num_cells % num_domains; ++i) {
    ++bounds[i + 1];
}
std::partial_sum(bounds.begin(), bounds.end(), bounds.begin());
```

Reply:
Ok I am convinced.

Reply:
Done.